src/wallet/spend.cpp

```diff
@@ -34,14 +34,60 @@ using interfaces::FoundBlock;
+    // Use max sig if watch only inputs were used or if this particular input is an external input
+    // to ensure a sufficient fee is attained for the requested feerate.
+    return coin_control && (coin_control->fAllowWatchOnly || (txin && coin_control->IsExternalSelected(txin->prevout)));
+}
+
+/** Get the size of an input (in witness units) once it's signed.
+ *
+ * @param desc The output script descriptor of the coin spent by this input.
+ * @param txin Optionally the txin to estimate the size of. Used to determine the size of ECDSA signatures.
+ * @param coin_control Information about the context to determine the size of ECDSA signatures.
+ * @param tx_is_segwit Whether the transaction has at least a single input spending a segwit coin.
+ * @param can_grind_r Whether the signer will be able to grind the R of the signature.
+ */
+static std::optional<int64_t> MaxInputWeight(const Descriptor& desc, const std::optional<CTxIn>& txin,
+                                             const CCoinControl* coin_control, const bool tx_is_segwit,
+                                             const bool can_grind_r) {
+    if (const auto sat_weight = desc.MaxSatisfactionWeight(!can_grind_r || UseMaxSig(txin, coin_control))) {
+        const bool is_segwit = IsSegwit(desc);
+        // Account for the size of the scriptsig and the number of elements on the witness stack. Note
+        // that if any input in the transaction is spending a witness program, we need to specify the
+        // witness stack size for every input regardless of whether it is segwit itself.
+        // NOTE: this also works in case of mixed scriptsig-and-witness such as in p2sh-wrapped segwit v0
+        // outputs. In this case the size of the scriptsig length will always be one (since the redeemScript
+        // is always a push of the witness program in this case, which is smaller than 253 bytes).
+        const int64_t scriptsig_len = is_segwit ? 1 : GetSizeOfCompactSize(*sat_weight / WITNESS_SCALE_FACTOR);
+        // FIXME: use the real number of stack elements instead of the "1" placeholder.
+        const int64_t witstack_len = is_segwit ? GetSizeOfCompactSize(1) : (tx_is_segwit ? 1 : 0);
+        // previous txid + previous vout + sequence + scriptsig len + witstack size + scriptsig or witness
+        // NOTE: sat_weight already accounts for the witness discount accordingly.
+        return (32 + 4 + 4 + scriptsig_len) * WITNESS_SCALE_FACTOR + witstack_len + *sat_weight;
+    }
+
+    return {};
+}
+
 int CalculateMaximumSignedInputSize(const CTxOut& txout, const COutPoint outpoint, const SigningProvider* provider, bool can_grind_r, const CCoinControl* coin_control)
 {
-    CMutableTransaction txn;
-    txn.vin.push_back(CTxIn(outpoint));
-    if (!provider || !DummySignInput(*provider, txn.vin[0], txout, can_grind_r, coin_control)) {
-        return -1;
+    if (!provider) return -1;
+
+    if (const auto desc = InferDescriptor(txout.scriptPubKey, *provider)) {
```

Comment:
This wouldn't miss any case that were handled previously as far as i can tell. `DummySignInput` would previously return `false` on a P2PKH for which we don't know the public key. However there may be ways to improve upon the current situation, for instance we could set `MaxInputWeight` for an `addr()`/`raw()` descriptor that represents a P2(W)PKH, since we know it? But i'd prefer to first get this switch to using descriptors done with no regression and leave improvements for a follow-up, if reviewers don't mind.